File: backend/services/class_service.py
from extensions import db
from repositories.class_repository import ClassRepository, ClassExamRepository, AnnouncementRepository
from repositories.lesson_repository import LessonRepository
from repositories.exam_repository import ExamRepository
from errors import NotFoundError, ConflictError

class_repo    = ClassRepository()
ce_repo       = ClassExamRepository()
ann_repo      = AnnouncementRepository()
lesson_repo   = LessonRepository()
exam_repo     = ExamRepository()


def get_teacher_classes(teacher_id: int, page: int = 1, limit: int = 12) -> dict:
    result = class_repo.find_by_teacher_with_stats(teacher_id, page, limit)
    items = []
    for r in result["items"]:
        total   = r["totalAnswers"] or 0
        correct = r["correctAnswers"] or 0
        avg_score = round(correct / total * 10, 2) if total else None
        pass_rate = round(correct / total * 100, 2) if total else 0
        status    = "good" if pass_rate >= 70 else "warning" if pass_rate >= 50 else "bad"
        items.append({
            "classId": r["classId"], "className": r["className"],
            "totalStudents": r["totalStudents"], "avgScore": avg_score,
            "passRate": pass_rate, "status": status,
        })
    return {"items": items, "total": result["total"], "page": result["page"], "pages": result["pages"]}


def get_class_detail(class_id: int, student_page: int = 1, student_limit: int = 15) -> dict:
    cls = class_repo.find_by_id(class_id)
    if not cls:
        raise NotFoundError("Lớp không tồn tại")
    paged     = class_repo.get_students_with_avg(class_id, student_page, student_limit)
    class_avg = class_repo.get_class_avg(class_id)
    return {
        "classId": cls.id, "className": cls.class_name,
        "totalStudents": paged["total"],
        "students": paged["items"],
        "studentPage": paged["page"],
        "studentPages": paged["pages"],
        "classAvg": class_avg,
    }


def add_class(teacher_id: int, class_name: str) -> None:
    class_repo.create(teacher_id, class_name)
    db.session.commit()


def update_class(class_id: int, teacher_id: int, class_name: str) -> None:
    cls = class_repo.find_by_id_and_teacher(class_id, teacher_id)
    if not cls:
        raise NotFoundError("Lớp không tồn tại")
    class_repo.update(cls, class_name)
    db.session.commit()


def delete_class(class_id: int, teacher_id: int) -> None:
    cls = class_repo.find_by_id_and_teacher(class_id, teacher_id)
    if not cls:
        raise NotFoundError("Lớp không tồn tại")
    class_repo.delete(cls)
    db.session.commit()


def get_class_exams(class_id: int, teacher_id: int) -> list:
    cls = class_repo.find_by_id_and_teacher(class_id, teacher_id)
    if not cls:
        raise NotFoundError("Lớp không tồn tại")
    rows = ce_repo.find_by_class_with_stats(class_id)
    for r in rows:
        r["deadline"]    = str(r["deadline"])    if r["deadline"]    else None
        r["assigned_at"] = str(r["assigned_at"]) if r["assigned_at"] else None
        r["open_time"]   = str(r["open_time"])   if r["open_time"]   else None
        r["time_limit"]  = int(r["time_limit"])  if r["time_limit"] is not None else None
    return rows


def assign_exam(class_id: int, teacher_id: int, exam_id: int, deadline, open_time, time_limit: int) -> None:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    try:
        ce_repo.assign(class_id, exam_id, deadline, open_time, time_limit)
        db.session.commit()
    except Exception as e:
        db.session.rollback()
        if "Duplicate" in str(e) or "uq_class_exam" in str(e):
            raise ConflictError(" Bài tập đã được giao cho lớp này")
        raise


def update_assignment(class_id: int, teacher_id: int, exam_id: int, deadline, open_time, time_limit: int) -> None:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    ce = ce_repo.find_by_class_and_exam(class_id, exam_id)
    if not ce:
        raise NotFoundError(" Bài tập chưa được giao cho lớp này")
    ce_repo.update(ce, deadline, open_time, time_limit)
    db.session.commit()


def unassign_exam(class_id: int, teacher_id: int, exam_id: int) -> None:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    ce = ce_repo.find_by_class_and_exam(class_id, exam_id)
    if not ce:
        raise NotFoundError(" Bài tập chưa được giao cho lớp này")
    ce_repo.unassign(ce)
    db.session.commit()


def get_students_with_scores(class_id: int) -> list:
    cls = class_repo.find_by_id(class_id)
    if not cls:
        raise NotFoundError("Lớp không tồn tại")
    return class_repo.get_students_with_scores(class_id)


def get_students_for_export(class_id: int, teacher_id: int) -> tuple:
    cls = class_repo.find_by_id_and_teacher(class_id, teacher_id)
    if not cls:
        raise NotFoundError("Lớp không tồn tại")
    return cls.class_name, class_repo.get_students_for_export(class_id)


def get_announcements(class_id: int) -> list:
    anns = ann_repo.find_by_class(class_id)
    return [{"id": a.id, "title": a.title, "content": a.content, "created_at": str(a.created_at) if a.created_at else None} for a in anns]


def create_announcement(class_id: int, teacher_id: int, title: str, content: str) -> int:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    ann = ann_repo.create(class_id, teacher_id, title, content)
    db.session.commit()
    return ann.id


def delete_announcement(ann_id: int, teacher_id: int) -> None:
    ann = ann_repo.find_by_id_and_teacher(ann_id, teacher_id)
    if not ann:
        raise NotFoundError("Thông báo không tồn tại")
    ann_repo.delete(ann)
    db.session.commit()

File: backend/src/services/class.service.js
import * as classRepo from '../repositories/class.repository.js';
import * as examRepo from '../repositories/exam.repository.js';
import { NotFoundError, ConflictError } from '../utils/error.utils.js';
import { formatDate, formatDateTime } from '../utils/date.utils.js';

export const getTeacherClasses = async (teacherId, page = 1, limit = 12) => {
  const result = await classRepo.findByTeacherWithStats(teacherId, page, limit);
  const items = result.items.map(r => {
    const total = r.totalAnswers || 0;
    const correct = r.correctAnswers || 0;
    const avgScore = total ? parseFloat((correct / total * 10).toFixed(2)) : null;
    const passRate = total ? parseFloat((correct / total * 100).toFixed(2)) : 0;
    const status = passRate >= 70 ? 'good' : passRate >= 50 ? 'warning' : 'bad';
    return {
      classId: r.classId, className: r.className,
      totalStudents: r.totalStudents, avgScore, passRate, status,
    };
  });
  return { items, total: result.total, page: result.page, pages: result.pages };
};

export const getClassDetail = async (classId, studentPage = 1, studentLimit = 15) => {
  const cls = await classRepo.findById(classId);
  if (!cls) throw new NotFoundError('Lớp không tồn tại');
  const paged = await classRepo.getStudentsWithAvg(classId, studentPage, studentLimit);
  const classAvg = await classRepo.getClassAvg(classId);
  return {
    classId: cls.id, className: cls.class_name,
    totalStudents: paged.total,
    students: paged.items.map(s => ({
      id: s.id,
      username: s.username,
      fullName: s.full_name,
      dob: formatDate(s.dob),
      avgScore: s.avg_score != null ? Number(s.avg_score) : null,
    })),
    studentPage: paged.page,
    studentPages: paged.pages,
    classAvg,
  };
};

export const addClass = async (teacherId, className) => {
  await classRepo.createClass(teacherId, className);
};

export const updateClass = async (classId, teacherId, className) => {
  const cls = await classRepo.findByIdAndTeacher(classId, teacherId);
  if (!cls) throw new NotFoundError('Lớp không tồn tại');
  await classRepo.updateClass(classId, className);
};

export const deleteClass = async (classId, teacherId) => {
  const cls = await classRepo.findByIdAndTeacher(classId, teacherId);
  if (!cls) throw new NotFoundError('Lớp không tồn tại');
  await classRepo.deleteClass(classId);
};

export const getClassExams = async (classId, teacherId) => {
  const cls = await classRepo.findByIdAndTeacher(classId, teacherId);
  if (!cls) throw new NotFoundError('Lớp không tồn tại');
  const rows = await classRepo.findClassExamsByClass(classId);
  return rows.map(r => ({
    ...r,
    deadline: formatDateTime(r.deadline),
    assigned_at: formatDateTime(r.assigned_at),
    open_time: formatDateTime(r.open_time),
    time_limit: r.time_limit != null ? parseInt(r.time_limit) : null,
  }));
};

export const assignExam = async (classId, teacherId, examId, deadline, openTime, timeLimit) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');
  const exam = await examRepo.findById(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  try {
    await classRepo.assignExam(classId, examId, deadline, openTime, timeLimit);
  } catch (e) {
    if (String(e).includes('Duplicate') || String(e).includes('uq_class_exam')) {
      throw new ConflictError('Bài tập đã được giao cho lớp này');
    }
    throw e;
  }
};

export const updateAssignment = async (classId, teacherId, examId, deadline, openTime, timeLimit) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');
  const ce = await classRepo.findClassExamByClassAndExam(classId, examId);
  if (!ce) throw new NotFoundError('Bài tập chưa được giao cho lớp này');
  await classRepo.updateExamAssignment(classId, examId, deadline, openTime, timeLimit);
};

export const unassignExam = async (classId, teacherId, examId) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');
  const ce = await classRepo.findClassExamByClassAndExam(classId, examId);
  if (!ce) throw new NotFoundError('Bài tập chưa được giao cho lớp này');
  await classRepo.unassignExam(classId, examId);
};

export const getStudentsWithScores = async (classId) => {
  const cls = await classRepo.findById(classId);
  if (!cls) throw new NotFoundError('Lớp không tồn tại');
  return classRepo.getStudentsWithScores(classId);
};

export const getStudentsForExport = async (classId, teacherId) => {
  const cls = await classRepo.findByIdAndTeacher(classId, teacherId);
  if (!cls) throw new NotFoundError('Lớp không tồn tại');
  return { className: cls.class_name, students: await classRepo.getStudentsForExport(classId) };
};

export const getAnnouncements = async (classId) => {
  const rows = await classRepo.findAnnouncementsByClass(classId);
  return rows.map(a => ({
    id: a.id, title: a.title, content: a.content,
    created_at: formatDate(a.created_at),
  }));
};

export const createAnnouncement = async (classId, teacherId, title, content) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');
  return classRepo.createAnnouncement(classId, teacherId, title, content);
};

export const deleteAnnouncement = async (annId, teacherId) => {
  const ann = await classRepo.findAnnouncementByIdAndTeacher(annId, teacherId);
  if (!ann) throw new NotFoundError('Thông báo không tồn tại');
  await classRepo.deleteAnnouncement(annId);
};
